### ISSUE TITLE
Issue #4: Implement G-Estimation of Structural Nested Models

### ISSUE THREAD

**Issue by datablogin** [datablogin/causal-inference-marketing#35]
## Summary
Implement G-estimation for structural nested models, providing a robust method for causal inference that is less sensitive to model misspecification than G-computation or IPW alone.

## Background
G-estimation is a method for estimating parameters of structural nested models that provides consistent estimates even when the outcome model is misspecified, as long as the treatment model is correct. This complements our existing doubly robust AIPW estimator with an alternative robust approach.

### Core Implementation
- [ ] Create `GEstimationEstimator` class inheriting from `BaseEstimator`
- [ ] Implement structural nested mean models (SNMMs)
- [ ] Parameter estimation via optimization (finding parameters where treatment effect is zero)
- [ ] Support for binary and continuous treatments
- [ ] Integration with existing treatment and outcome models

### Structural Nested Models
- [ ] **Linear SNMMs**: Additive treatment effects
- [ ] **Multiplicative SNMMs**: Relative treatment effects  
- [ ] **General SNMMs**: Flexible functional forms
- [ ] **Time-varying treatments**: Extension to longitudinal data (future enhancement)

### Optimization Methods
- [ ] **Grid Search**: Systematic parameter space exploration
- [ ] **Root Finding**: Efficient single-parameter estimation
- [ ] **Gradient-based**: For multi-parameter models
- [ ] **Confidence Intervals**: Via inversion of test statistics

### Enhanced Features
- [ ] Rank preservation tests for model validity
- [ ] Sensitivity analysis for model assumptions
- [ ] Comparison with G-computation and IPW estimates
- [ ] Bootstrap confidence intervals with G-estimation
- [ ] Diagnostic plots and model checking tools

### API Design
```python
from causal_inference.estimators.g_estimation import GEstimationEstimator

# Basic G-estimation
estimator = GEstimationEstimator(
    structural_model="linear",  # or "multiplicative", "general"
    treatment_model="logistic",
    optimization_method="grid_search",
    parameter_range=(-10, 10),
    n_grid_points=1000,
    bootstrap_samples=1000
)

estimator.fit(treatment, outcome, covariates)
effect = estimator.estimate_ate()

# Get optimization details
optim_results = estimator.get_optimization_results()
rank_test = estimator.rank_preservation_test()

# Compare with other methods
comparison = estimator.compare_with_other_methods(
    methods=['g_computation', 'ipw', 'aipw']
)
```

### Statistical Performance KPIs
1. **Parameter Recovery**: Recover true parameters within 5% in simulation studies
2. **Robustness**: Maintain performance when outcome model is misspecified
3. **Optimization Convergence**: Find optimal parameters in 95% of cases
4. **Rank Preservation**: Rank preservation test has appropriate size and power

### Computational Performance KPIs
1. **Optimization Speed**: Complete parameter estimation on NHEFS dataset in <30 seconds
2. **Grid Search Efficiency**: Coarse-to-fine grid search reduces computation by 50%
3. **Bootstrap Performance**: 1000 bootstrap samples complete in <10 minutes
4. **Memory Usage**: Peak memory <500MB during optimization

### Quality KPIs
1. **Convergence Rate**: Optimization converges in ≥95% of realistic scenarios
2. **CI Coverage**: Bootstrap confidence intervals have 94-96% coverage
3. **Sensitivity**: Reasonable performance across different structural model specifications
4. **Diagnostics**: Rank preservation test correctly identifies model violations

### Test Case 1: Linear Structural Nested Mean Model
```python
# Load NHEFS data for smoking cessation analysis
loader = NHEFSDataLoader()
data = loader.load_processed_data()

treatment = TreatmentData(values=data['qsmk'], treatment_type="binary")
outcome = OutcomeData(values=data['wt82_71'], outcome_type="continuous")
covariates = CovariateData(
    values=data[['age', 'sex', 'race', 'education', 'smokeintensity', 'smokeyrs']],
    names=['age', 'sex', 'race', 'education', 'smokeintensity', 'smokeyrs']
)

# G-estimation with linear SNMM
estimator = GEstimationEstimator(
    structural_model="linear",
    treatment_model="logistic",
    optimization_method="grid_search",
    parameter_range=(-5, 10),
    n_grid_points=500
)

estimator.fit(treatment, outcome, covariates)
effect = estimator.estimate_ate()

# Expected results (benchmark from reference implementation)
expected_ate_range = (2.0, 5.0)  # kg weight change
assert expected_ate_range[0] <= effect.ate <= expected_ate_range[1]

# Check optimization convergence
optim_results = estimator.get_optimization_results()
assert optim_results['converged'] == True
assert optim_results['objective_value'] < 0.01  # Should be close to zero
```

### Test Case 2: Robustness to Model Misspecification
```python
# Test with deliberately misspecified outcome model
misspec_covariates = CovariateData(
    values=data[['age', 'sex']],  # Deliberately omit important confounders
    names=['age', 'sex']
)

# G-estimation should still work if treatment model is correct
estimator_robust = GEstimationEstimator(
    structural_model="linear",
    treatment_model="logistic"  # Include all confounders in treatment model
)

# Fit with full covariate set for treatment model but limited for outcome
estimator_robust.fit(treatment, outcome, covariates)
effect_robust = estimator_robust.estimate_ate()

# Compare with G-computation using misspecified model
g_comp_estimator = GComputationEstimator()
g_comp_estimator.fit(treatment, outcome, misspec_covariates)
effect_g_comp = g_comp_estimator.estimate_ate()

# G-estimation should be more robust (closer to true effect)
```

### Test Case 3: Rank Preservation Test
```python
# Test rank preservation assumption
estimator = GEstimationEstimator(structural_model="linear")
estimator.fit(treatment, outcome, covariates)

# Perform rank preservation test
rank_test = estimator.rank_preservation_test()
assert 'test_statistic' in rank_test
assert 'p_value' in rank_test
assert 'conclusion' in rank_test

# Should not reject rank preservation for well-specified model
assert rank_test['p_value'] > 0.05  # Expect not to reject
```

### Test Case 4: Multi-parameter G-estimation
```python
# Test with interaction terms in structural model
estimator_interact = GEstimationEstimator(
    structural_model="linear_with_interactions",
    covariates_for_interaction=['age', 'sex'],
    optimization_method="gradient_based"
)

estimator_interact.fit(treatment, outcome, covariates)
effect_interact = estimator_interact.estimate_ate()

# Should estimate main effect and interaction effects
params = estimator_interact.get_estimated_parameters()
assert 'main_effect' in params
assert 'age_interaction' in params  
assert 'sex_interaction' in params
```

### Test Case 5: Comparison with Other Methods
```python
# Compare G-estimation with other estimators
comparison = estimator.compare_with_other_methods(
    treatment=treatment,
    outcome=outcome, 
    covariates=covariates,
    methods=['g_computation', 'ipw', 'aipw']
)

# All methods should give reasonably similar results
estimates = [comparison[method]['ate'] for method in comparison.keys()]
assert max(estimates) - min(estimates) < 2.0  # Should be within 2kg

# G-estimation should have reasonable precision
assert comparison['g_estimation']['ate_se'] < 1.0
```

## Implementation References
- Chapter 14 "G-estimation of Structural Nested Models" from "Causal Inference: What If"
- Robins, J. M. (1994). "Correcting for non-compliance in randomized trials using structural nested mean models"
- Vansteelandt, S. & Joffe, M. (2014). "Structural nested models and G-estimation"
- jrfiedler/causal_inference_python_code (Chapter 14 notebook)

## Definition of Done
- [ ] G-estimation implemented for linear structural nested models
- [ ] Optimization algorithms working correctly and converging
- [ ] Rank preservation test implemented and validated
- [ ] Bootstrap confidence intervals integrated
- [ ] Comparison functionality with other estimators
- [ ] Performance benchmarks meet specified KPIs
- [ ] Robustness demonstrated in misspecification scenarios

## Acceptance Criteria
1. G-estimation recovers parameters within 5% accuracy on NHEFS dataset
2. Optimization converges in ≥95% of realistic test cases
3. Method remains robust when outcome model is deliberately misspecified
4. Rank preservation test correctly identifies assumption violations
5. Bootstrap confidence intervals show appropriate coverage (94-96%)
6. Performance meets computational KPIs for optimization and bootstrap
7. Integration with existing architecture maintains API consistency

## Priority: P3 (Medium)
This provides an important alternative robust method that complements our existing doubly robust AIPW estimator, particularly valuable when model misspecification is a concern.
